x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/Rate.java

```diff
@@ -0,0 +1,162 @@
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamOutput;
+import org.elasticsearch.xpack.esql.planner.ToAggregator;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypes;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.List;
+
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.esql.core.expression.TypeResolutions.isType;
+
+public class Rate extends AggregateFunction implements OptionalArgument, ToAggregator {
+    private static final TimeValue DEFAULT_UNIT = TimeValue.timeValueSeconds(1);
+
+    private final Expression timestamp;
+    private final Expression unit;
+
+    @FunctionInfo(
+        returnType = { "double" },
+        description = "compute the rate of a counter field. Available in METRICS command only",
+        isAggregation = true
+    )
+
+    public Rate(
+        Source source,
+        @Param(name = "field", type = { "counter_long|counter_integer|counter_double" }, description = "counter field") Expression field,
+        Expression timestamp,
+        @Param(optional = true, name = "unit", type = { "time_duration" }, description = "the unit") Expression unit
+    ) {
+        super(source, field, unit != null ? List.of(timestamp, unit) : List.of(timestamp));
+        this.timestamp = timestamp;
+        this.unit = unit;
+    }
+
+    public static Rate withUnresolvedTimestamp(Source source, Expression field, Expression unit) {
+        return new Rate(source, field, new UnresolvedAttribute(source, "@timestamp"), unit);
+    }
+
+    public static Rate readRate(PlanStreamInput in) throws IOException {
+        Source source = Source.readFrom(in);
+        Expression field = in.readExpression();
+        Expression timestamp = in.readExpression();
+        Expression unit = in.readBoolean() ? in.readExpression() : null;
+        return new Rate(source, field, timestamp, unit);
+    }
+
+    public static void writeRate(PlanStreamOutput out, Rate rate) throws IOException {
+        rate.source().writeTo(out);
+        out.writeExpression(rate.field());
+        out.writeExpression(rate.timestamp);
+        if (rate.unit != null) {
+            out.writeBoolean(true);
+            out.writeExpression(rate.unit);
+        } else {
+            out.writeBoolean(false);
+        }
+    }
+
+    @Override
+    protected NodeInfo<Rate> info() {
+        return NodeInfo.create(this, Rate::new, field(), timestamp, unit);
+    }
+
+    @Override
+    public Rate replaceChildren(List<Expression> newChildren) {
+        final Expression unit = newChildren.size() >= 3 ? newChildren.get(2) : null;
```

Comment:
Could the size of `newChildren` be 4 here? Sounds like a bug if that is the case. Maybe we should just check for whether the size of the list is 2 or 3 and otherwise throw an exception?